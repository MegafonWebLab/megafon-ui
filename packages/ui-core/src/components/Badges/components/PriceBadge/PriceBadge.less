@import (reference) '~styles/base.less';

@b: e('.mfui-price-badge');

@{b} {
    .smallFont();

    display: flex;
    align-items: center;

    padding: 2px;
    padding-right: 8px;
    width: fit-content;
    height: fit-content;

    border-radius: 8px;

    &_adaptive {
        @media @desktop {
            .commonFont();

            padding: 4px;
            padding-right: 8px;
            min-height: 32px;

            border-radius: 12px;
        }
    }
    
    &_theme_grey {
        background-color: var(--spbSky0);

        @{b}__text {
            color: var(--spbSky3);
        }

        @{b}__icon {
            fill: var(--spbSky3);
        }
    }

    &_theme_orange {
        background-color: var(--137C20);

        @{b}__text {
            color: var(--137C);
        }

        @{b}__icon {
            fill: var(--137C);
        }
    }

    &_theme_green {
        background-color: var(--brandGreen20);

        @{b}__text {
            color: var(--brandGreen);
        }

        @{b}__icon {
            fill: var(--brandGreen);
        }
    }

    &_theme_red {
        background-color: var(--fury20);

        @{b}__text {
            color: var(--fury);
        }

        @{b}__icon {
            fill: var(--fury);
        }
    }

    &_size_big {
        font-size: 15px;
        line-height: 18px;
        padding: 12px 12px 12px 4px;
    }

    &__text {
        .fontFamily();

        margin-left: 2px;
    }

    &__icon-container {
        display: flex;
    }

    &__icon {
        height: 20px;
        width: 20px;
        min-width: 20px;
    }

    &_size_big & {
        &__icon {
            height: 32px;
            width: 32px;
            min-width: 32px;    
        }

        &__sub-title {
            font-size: 12px;
            line-height: 18px;
            margin-top: 2px;
        }
    }
}
