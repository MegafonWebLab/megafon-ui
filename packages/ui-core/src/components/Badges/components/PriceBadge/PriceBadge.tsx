import * as React from 'react';
import { cnCreate, filterDataAttrs } from '@megafon/ui-helpers';
import AttentionIcon from '@megafon/ui-icons/system-16-attention_16.svg';
import CheckIcon from '@megafon/ui-icons/system-16-check_invert_16.svg';
import PriceIcon from '@megafon/ui-icons/system-16-price_16.svg';
import TimerIcon from '@megafon/ui-icons/system-16-timer_16.svg';
import AttentionIconBig from '@megafon/ui-icons/system-24-attention_24.svg';
import CheckIconBig from '@megafon/ui-icons/system-24-check_invert_24.svg';
import PriceIconBig from '@megafon/ui-icons/system-24-price_24.svg';
import TimerIconBig from '@megafon/ui-icons/system-24-timer_24.svg';
import * as PropTypes from 'prop-types';
import './PriceBadge.less';

export const PriceBadgeTheme = {
    RED: 'red',
    GREY: 'grey',
    GREEN: 'green',
    ORANGE: 'orange',
} as const;

export const PriceBadgeIcon = {
    TIMER: 'timer',
    PRICE: 'price',
    CHECK: 'check',
    ATTENTION: 'attention',
} as const;

export const PriseBadgeSize = {
    SMALL: 'small',
    BIG: 'big',
} as const;

type PriceBadgeThemeType = typeof PriceBadgeTheme[keyof typeof PriceBadgeTheme];
type PriceBadgeIconType = typeof PriceBadgeIcon[keyof typeof PriceBadgeIcon];
type PriseBadgeSizeType = typeof PriseBadgeSize[keyof typeof PriseBadgeSize];

const getPriceBadgeIcon = (iconType: PriceBadgeIconType, size: PriseBadgeSizeType) => {
    const isBigIcon = size === PriseBadgeSize.BIG;

    switch (true) {
        case iconType === PriceBadgeIcon.TIMER && isBigIcon:
            return TimerIconBig;
        case iconType === PriceBadgeIcon.PRICE && isBigIcon:
            return PriceIconBig;
        case iconType === PriceBadgeIcon.CHECK && isBigIcon:
            return CheckIconBig;
        case iconType === PriceBadgeIcon.ATTENTION && isBigIcon:
            return AttentionIconBig;
        case iconType === PriceBadgeIcon.PRICE:
            return PriceIcon;
        case iconType === PriceBadgeIcon.CHECK:
            return CheckIcon;
        case iconType === PriceBadgeIcon.ATTENTION:
            return AttentionIcon;
        default:
            return TimerIcon;
    }
};

export interface IPriceBadgeProps {
    /** Адаптивный режим */
    isAdaptive?: boolean;
    /** Тип иконки */
    iconType?: PriceBadgeIconType;
    /** Цветовая тема */
    theme?: PriceBadgeThemeType;
    /** Дополнительный класс корневого элемента */
    className?: string;
    /** Размер бейджа */
    size?: PriseBadgeSizeType;
    /** Дополнительный текст (только для size="big") */
    subTitle?: string;
    /** Дополнительные data-атрибуты к внутренним элементам */
    dataAttrs?: {
        root?: Record<string, string>;
    };
    children: JSX.Element[] | Element[] | JSX.Element | Element | string;
}

const cn = cnCreate('mfui-price-badge');
const PriceBadge: React.FC<IPriceBadgeProps> = ({
    iconType = 'timer',
    isAdaptive = false,
    theme = 'grey',
    size = 'small',
    subTitle,
    className,
    dataAttrs,
    children,
}) => {
    const Icon = getPriceBadgeIcon(iconType, size);
    const showSubTitle = size === 'big' && !!subTitle;

    return (
        <div {...filterDataAttrs(dataAttrs?.root)} className={cn({ theme, adaptive: isAdaptive, size }, className)}>
            <div className={cn('icon-container')}>
                <Icon className={cn('icon')} />
            </div>
            <div className={cn('text')}>
                <div>{children}</div>
                {showSubTitle && <div className={cn('sub-title')}>{subTitle}</div>}
            </div>
        </div>
    );
};

PriceBadge.propTypes = {
    isAdaptive: PropTypes.bool,
    iconType: PropTypes.oneOf(Object.values(PriceBadgeIcon)),
    theme: PropTypes.oneOf(Object.values(PriceBadgeTheme)),
    size: PropTypes.oneOf(Object.values(PriseBadgeSize)),
    subTitle: PropTypes.string,
    className: PropTypes.string,
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
    }),
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.element.isRequired),
        PropTypes.element,
        PropTypes.string,
    ]).isRequired,
};

export default PriceBadge;
