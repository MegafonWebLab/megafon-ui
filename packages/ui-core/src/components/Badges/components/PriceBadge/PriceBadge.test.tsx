import * as React from 'react';
import { render } from '@testing-library/react';
import PriceBadge, { IPriceBadgeProps } from './PriceBadge';

const dataAttrs: IPriceBadgeProps['dataAttrs'] = {
    root: {
        'data-testid': 'root',
    },
};

describe('<PriceBadge />', () => {
    it('should render PriceBadge', () => {
        const { container } = render(<PriceBadge>text</PriceBadge>);

        expect(container).toMatchSnapshot();
    });

    it('should render with className', () => {
        const { getByTestId } = render(
            <PriceBadge dataAttrs={dataAttrs} className="custom-class">
                text
            </PriceBadge>,
        );

        expect(getByTestId('root')).toHaveClass('custom-class');
    });

    it('should render with dataAttrs', () => {
        const { queryByTestId } = render(<PriceBadge dataAttrs={dataAttrs}>text</PriceBadge>);

        expect(queryByTestId('root')).toBeTruthy();
    });

    it('should render with theme different from default', () => {
        const { getByTestId } = render(
            <PriceBadge dataAttrs={dataAttrs} theme="orange">
                text
            </PriceBadge>,
        );

        expect(getByTestId('root')).toHaveClass('mfui-price-badge_theme_orange');
    });

    it('should render when isAdaptive is true', () => {
        const { getByTestId } = render(
            <PriceBadge dataAttrs={dataAttrs} isAdaptive>
                text
            </PriceBadge>,
        );

        expect(getByTestId('root')).toHaveClass('mfui-price-badge_adaptive');
    });
});
