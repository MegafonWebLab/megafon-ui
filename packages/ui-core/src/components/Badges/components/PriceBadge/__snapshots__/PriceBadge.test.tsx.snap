// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PriceBadge /> should render PriceBadge 1`] = `
<div>
  <div
    class="mfui-price-badge mfui-price-badge_theme_grey"
  >
    <div
      class="mfui-price-badge__icon-container"
    >
      <test-file-stub
        classname="mfui-price-badge__icon"
      />
    </div>
    <span
      class="mfui-price-badge__text"
    >
      text
    </span>
  </div>
</div>
`;
